### PR TITLE
Use persistent storage by default for ES data

### PR DESCRIPTION
**Pull request by nkvoll**
ℹ️  Builds on top of https://github.com/elastic/cloud-on-k8s/pull/1024, only the last commit is relevant for this specific PR.

By default they will get a 1Gi volume, which is chosen because matches roughly 1:1 vs the default heap.

Users can opt out of this behavior by specifying the data volume in the Elasticsearch resource directly

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: elasticsearch-sample
spec:
  version: "7.1.0"
  nodes:
  - nodeCount: 1
    podTemplate:
      spec:
        volumes:
        - name: elasticsearch-data
          emptyDir:
            # specifying "medium" is not necessary, but shows how to use e.g tmpfs on linux
            medium: Memory
```

Closes: https://github.com/elastic/cloud-on-k8s/issues/913

### CODE REVIEW

File: operators/pkg/controller/elasticsearch/pod/pod.go [hunk 1]
```diff
@@ -29,6 +31,25 @@ var (
 		{Name: "transport", ContainerPort: network.TransportPort, Protocol: corev1.ProtocolTCP},
 		{Name: "process-manager", ContainerPort: processmanager.DefaultPort, Protocol: corev1.ProtocolTCP},
 	}
+
+	// DefaultVolumeClaimsTemplates is the default volume claim templates for Elasticsearch pods
+	DefaultVolumeClaimsTemplates = []corev1.PersistentVolumeClaim{
+		{
+			ObjectMeta: v1.ObjectMeta{
+				Name: "elasticsearch-data",
```
Comment (sebgl) [inline operators/pkg/controller/elasticsearch/pod/pod.go]: Should reuse a const?

File: operators/pkg/controller/elasticsearch/pod/pod.go [hunk 2]
```diff
@@ -29,6 +31,25 @@ var (
...
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{
+					corev1.ReadWriteOnce,
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: resource.MustParse("1Gi"),
```
Comment (sebgl) [inline operators/pkg/controller/elasticsearch/pod/pod.go]: I'm wondering if this is big enough as a default. Could we get into a situation where many users start with a simple copy/pasted spec, get their 1Gi storage full and have basically lost their data at this point?
I'm seeing [Helm Charts default to 30Gi](https://github.com/elastic/helm-charts/blob/51ab321e5640e282392087536c554e4bd4239fd2/elasticsearch/values.yaml#L76), maybe we should align here?
Not sure that would fit Minikube very well though :/

File: operators/pkg/controller/elasticsearch/version/common.go [hunk 3]
```diff
@@ -70,6 +79,21 @@ func NewExpectedPodSpecs(
...
+func shouldAddDefaultPVCToNodeSpec(defaultPVC corev1.PersistentVolumeClaim, node v1alpha1.NodeSpec) bool {
+	for _, volume := range node.PodTemplate.Spec.Volumes {
+		if volume.Name == defaultPVC.Name {
+			if volume.PersistentVolumeClaim == nil {
+				// the template contains a custom volume which is not a PVC, so we should not add this
+				// defaulted PVC template
+				return false
+			}
+		}
+	}
```
Comment (sebgl) [inline operators/pkg/controller/elasticsearch/version/common.go]: could this logic live within `defaults.AppendDefaultPVCs` directly?

File: operators/pkg/controller/common/defaults/pvc.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package defaults
+
+import v1 "k8s.io/api/core/v1"
```
Comment (thbkrkr) [inline operators/pkg/controller/common/defaults/pvc.go]: Nit: `v1` alias is useless here. We used to use `corev1`.

File: operators/pkg/controller/elasticsearch/pod/pod.go [hunk 5]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/settings"
 	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/volume"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/pod/pod.go]: `metav1` import alias?

File: operators/pkg/controller/common/defaults/pvc.go (hunk 4 above)
Comment (nkvoll) [reply operators/pkg/controller/common/defaults/pvc.go]: I know, but `make fmt` adds it back in again if I remove it :(

We're very inconsistent about `v1` vs `corev1` in the rest of the code base. I think we're going to need a cleanup-PR and a CI check to ensure consistency in the long term here.

File: operators/pkg/controller/elasticsearch/pod/pod.go (hunk 2 above)
Comment (nkvoll) [reply operators/pkg/controller/elasticsearch/pod/pod.go]: 1Gi works well with Minikube etc, and doesn't accidentally incur a large cost. It's probably best if users are conscious about this number in any production scenario.

File: operators/pkg/controller/common/defaults/pvc.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+// AppendDefaultPVCs appends defaults PVCs to a set of existing ones.
+//
+// The default not appended if:
```
Comment (sebgl) [inline operators/pkg/controller/common/defaults/pvc.go]: I think we may be missing a word in that sentence?

File: operators/pkg/controller/common/defaults/pvc.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+// - a Volume with a the same .name is found in podSpec.Volumes, and that volume is not a PVC volume
```
Comment (sebgl) [inline operators/pkg/controller/common/defaults/pvc.go]: `a the`

File: operators/pkg/controller/elasticsearch/volume/volume.go [hunk 8]
```diff
@@ -33,6 +33,9 @@ const (
 	UnicastHostsFile            = "unicast_hosts.txt"
 
 	ProcessManagerEmptyDirMountPath = "/mnt/elastic-internal/process-manager"
+
+	ElasticsearchDataVolumeName = "elasticsearch-data"
```
Comment (sebgl) [inline operators/pkg/controller/elasticsearch/volume/volume.go]: We should probably update all samples and the quickstart doc to use this name in the VolumeClaimTemplates section.

Comment (nkvoll) [reply operators/pkg/controller/elasticsearch/volume/volume.go]: Already updated that in the "normalization" PR.

**Comment by nkvoll** [comment]
Jenkins test this please

File: operators/pkg/controller/elasticsearch/pod/pod.go [hunk 9]
```diff
@@ -31,6 +32,25 @@ var (
 		{Name: "transport", ContainerPort: network.TransportPort, Protocol: corev1.ProtocolTCP},
 		{Name: "process-manager", ContainerPort: processmanager.DefaultPort, Protocol: corev1.ProtocolTCP},
 	}
+
+	// DefaultVolumeClaimsTemplates is the default volume claim templates for Elasticsearch pods
+	DefaultVolumeClaimsTemplates = []corev1.PersistentVolumeClaim{
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/pod/pod.go]: Nit: `DefaultVolumeClaimTemplates` with `Claim` in the singular?

File: operators/pkg/controller/common/defaults/pvc.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	defaults ...v1.PersistentVolumeClaim,
+) []v1.PersistentVolumeClaim {
+	// create a set of volume names that are not PVC-volumes for efficient testing
+	nonPVCvolumes := map[string]struct{}{}
+
+	for _, volume := range podSpec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			// this volume is not a PVC
+			nonPVCvolumes[volume.Name] = struct{}{}
+
```
Comment (thbkrkr) [inline operators/pkg/controller/common/defaults/pvc.go]: Unnecessary blank line.

**Pull request merged**